src/main/java/io/supertokens/bulkimport/BulkImportUserUtils.java

```diff
@@ -88,80 +104,123 @@ private static List<TotpDevice> getParsedTotpDevices(JsonObject userData, List<S
+            long timeJoinedInMSSinceEpoch = timeJoined != null ? timeJoined.longValue() : 0;
+
+            validateTenantId(main, appIdentifier, tenantId, recipeId, errors);
+
+            if ("emailpassword".equals(recipeId)) {
+                String email = parseAndValidateField(jsonLoginMethodObj, "email", ValueType.STRING, true, String.class, errors, " for an emailpassword recipe.");
+                String passwordHash = parseAndValidateField(jsonLoginMethodObj, "passwordHash", ValueType.STRING, true, String.class, errors, " for an emailpassword recipe.");
+                String hashingAlgorithm = parseAndValidateField(jsonLoginMethodObj, "hashingAlgorithm", ValueType.STRING, true, String.class, errors, " for an emailpassword recipe.");
+
+                email = email != null ? Utils.normaliseEmail(email) : null;
+                hashingAlgorithm = hashingAlgorithm != null ? hashingAlgorithm.trim().toUpperCase() : null;
+
+                EmailPasswordLoginMethod emailPasswordLoginMethod = new EmailPasswordLoginMethod(email, passwordHash, hashingAlgorithm);
+
+                validatePasswordHashingAlgorithm(main, appIdentifier, emailPasswordLoginMethod, errors);
+
+                loginMethods.add(new LoginMethod(tenantId, recipeId, isVerified, isPrimary, timeJoinedInMSSinceEpoch, emailPasswordLoginMethod, null, null));
+            } else if ("thirdparty".equals(recipeId)) {
+                String email = parseAndValidateField(jsonLoginMethodObj, "email", ValueType.STRING, true, String.class, errors, " for a thirdparty recipe.");
+                String thirdPartyId = parseAndValidateField(jsonLoginMethodObj, "thirdPartyId", ValueType.STRING, true, String.class, errors, " for a thirdparty recipe.");
+                String thirdPartyUserId = parseAndValidateField(jsonLoginMethodObj, "thirdPartyUserId", ValueType.STRING, true, String.class, errors, " for a thirdparty recipe.");
+
+                email = email != null ? Utils.normaliseEmail(email) : null;
+
+                ThirdPartyLoginMethod thirdPartyLoginMethod = new ThirdPartyLoginMethod(email, thirdPartyId, thirdPartyUserId);
+                loginMethods.add(new LoginMethod(tenantId, recipeId, isVerified, isPrimary, timeJoinedInMSSinceEpoch, null, thirdPartyLoginMethod, null));
+            } else if ("passwordless".equals(recipeId)) {
+                String email = parseAndValidateField(jsonLoginMethodObj, "email", ValueType.STRING, false, String.class, errors, " for a passwordless recipe.");
+                String phoneNumber = parseAndValidateField(jsonLoginMethodObj, "phoneNumber", ValueType.STRING, false, String.class, errors, " for a passwordless recipe.");
+
+                email = email != null ? Utils.normaliseEmail(email) : null;
+                phoneNumber = Utils.normalizeIfPhoneNumber(phoneNumber);
```

Comment:
what if phone number is null?

Reply:
`normalizeIfPhoneNumber` was checking for the null case but after refactoring, I am doing the check in the function as well.